### PR TITLE
perf(torch): skip the flatten walk for batch-size extraction in TorchTrainer

### LINKED ISSUE #23282: TorchTrainer takes the slow loss/batch-size path every step (and crashes GPU fit with a device bug)
On the PyTorch backend, every training/eval step redoes work that the loss and batch-size handling don't actually need to redo:

- `Loss.__call__` in `keras/src/losses/loss.py` always opens a `name_scope` and routes `y_true`/`y_pred` through `convert_to_tensor` (dtype standardization + device transfer), even when both are already plain `torch.Tensor`s on the same device and already match the loss's dtype -- the steady-state case for every step of `model.fit()`/`model.evaluate()`.
- `TorchTrainer.train_step`/`test_step` in `keras/src/backend/torch/trainer.py` extracts the batch size by calling `tree.flatten(x)` and reading the first leaf's shape, even when `x` is a single plain tensor -- `tree.flatten` is unconditionally walking a structure that's already a known single leaf in that case.

Neither check depends on anything that changes between steps, so both run their full general-case logic on every single call instead of only when the input actually needs it (dict/list inputs, mixed dtypes, cross-device tensors, sample weighting, or masking).

Fixed by #23187: adds a fast path in `Loss.__call__` gated on tensor type/dtype/device match (with a device guard added specifically to avoid a GPU `fit()` crash when the fast path skips `convert_to_tensor`'s device transfer), and a direct `x.shape[0]` read in `TorchTrainer` for plain-tensor inputs, falling back to the slow `tree.flatten` path for dict/list inputs.

Part of the per-call Python-dispatch overhead series in #22561.

### PR DESCRIPTION
**Pull request by pctablet505**
Closes #23282

## Summary

`TorchTrainer.train_step` and `test_step` derive the batch size for the loss tracker with `next(i for i in tree.flatten(x) if i is not None).shape[0]`. When `x` is a plain tensor — the common case — that builds a flattened list to read one shape. This reads `x.shape[0]` directly and falls back to the existing walk for dict and list inputs.

The change is confined to `keras/src/backend/torch/trainer.py`.

## Scope

This PR previously also added a fast path to `Loss.__call__`. That is removed. It required a `backend.backend() == "torch"` test inside backend-independent `keras/src/losses/loss.py`, and the guard it needed cannot be written backend-independently: it has to know that `y_pred` and `y_true` are already tensors on the same device, and on the torch backend `convert_to_tensor` *is* the device transfer, so there is no separate predicate to ask.

## Effect

Small. The flatten it avoids is one generator plus a list build per step, about 0.157 µs, roughly 0.023% of a step. It is a strict reduction in work on the common path rather than a measurable speed-up, so treat it as cleanup rather than optimization.

## Contributor Agreement

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

**Comment by pctablet505** [comment]
@gemini-code-assist review

File: keras/src/losses/loss_test.py [hunk 1]
```diff
@@ -1,22 +1,90 @@
...
+def _to_np(t):
+    """Convert tensor or array to numpy for numeric comparison."""
+    try:
+        import torch as _torch
+
+        if isinstance(t, _torch.Tensor):
+            return t.detach().cpu().numpy()
+    except ImportError:
+        pass
+    return np.asarray(t)
```
Comment (hertschuh) [inline keras/src/losses/loss_test.py:30]: Remove, that's already done by `assertAllClose`

File: keras/src/losses/loss.py [hunk 2]
```diff
@@ -54,6 +54,50 @@ def dtype(self):
...
+        # already match self.dtype, live on the same device, carry no input
+        # mask, and have no sample_weight. Any input that doesn't satisfy
+        # every condition (a numpy/list/dict y_true, a mismatched dtype or
+        # device, an existing mask) falls through to the slow path below,
+        # which converts and casts it correctly.
+        if (
+            sample_weight is None
+            and backend.backend() == "torch"
+            and not isinstance(y_pred, (dict, list, tuple))
+        ):
```
Comment (hertschuh) [inline keras/src/losses/loss.py:68]: `not isinstance(y_pred, (dict, list, tuple))`

This is not enough to determine if `y_pred` is not nested, there could be other structures, in particular registered structures.

File: keras/src/losses/loss.py [hunk 3]
```diff
@@ -54,6 +54,50 @@ def dtype(self):
...
+                # already skips the sample_weight/mask handling that
+                # reduce_weighted_values would otherwise do, so routing
+                # through it here is no slower than hand-rolling mean()/
+                # sum() directly, while staying dtype-exact with the slow
+                # path: divide_no_nan forces 64-bit compute dtypes down to
+                # their 32-bit equivalent on non-tensorflow backends, which
+                # a bare losses.mean() would not.
+                return reduce_values(
+                    losses, sample_weight=None, reduction=self.reduction
+                )
```
Comment (hertschuh) [inline keras/src/losses/loss.py:99]: We should not have torch specific code in backend independent code.

You could make this backend independent. But please move it inside the `name_scope`, because skipping the `name_scope` would be functionally different.

File: keras/src/losses/loss.py [hunk 4]
```diff
@@ -54,6 +54,50 @@ def dtype(self):
...
+                and isinstance(y_true, _torch.Tensor)
+                and y_pred.device == y_true.device
+                and backend.standardize_dtype(y_pred.dtype) == self.dtype
+                and backend.standardize_dtype(y_true.dtype) == self.dtype
+                and backend.get_keras_mask(y_pred) is None
+            ):
+                losses = self.call(y_true, y_pred)
+                mask = backend.get_keras_mask(losses)
+                if mask is not None:
+                    return reduce_weighted_values(
```
Comment (hertschuh) [inline keras/src/losses/loss.py:82]: How is this possible?

Does this mean `y_true` has a mask? (but `y_pred` doesn't?)

**Review by hertschuh — CHANGES_REQUESTED** [review]

File: keras/src/backend/torch/trainer.py [hunk 5]
```diff
@@ -108,11 +108,15 @@ def train_step(self, data):
         loss = self._compute_loss(
             x=x, y=y, y_pred=y_pred, sample_weight=sample_weight, training=True
         )
+        # Fast batch-size extraction: avoid tree.flatten for the common case
+        # where x is a plain tensor (dict/list inputs fall back to flatten).
```
Comment (hertschuh) [inline keras/src/backend/torch/trainer.py]: I agree with the code change, but the comment really only make sense within the context of this PR.

Also, it looks like the comment applies to `self._loss_tracker.update_state(` which is confusing.

Anyway, remove the comment. People can use "Blame" to retrieve the context.

File: keras/src/backend/torch/trainer.py [hunk 6]
```diff
@@ -148,11 +152,15 @@ def test_step(self, data):
         loss = self._compute_loss(
             x=x, y=y, y_pred=y_pred, sample_weight=sample_weight, training=False
         )
+        # Fast batch-size extraction: avoid tree.flatten for the common case
+        # where x is a plain tensor (dict/list inputs fall back to flatten).
```
Comment (hertschuh) [inline keras/src/backend/torch/trainer.py]: Remove comment.

**Review by hertschuh — CHANGES_REQUESTED** [review]
Thanks, can you remove the comments and mark as "Ready for Review"?

**Comment by pctablet505** [comment]
> Thanks, can you remove the comments and mark as "Ready for Review"?

It is marked ready for review. Did the rework as suggested.

**Review by hertschuh — APPROVED** [review]
Thanks for the improvement!
